HTTP_API.md

```diff
@@ -0,0 +1,952 @@
+    ```
+
+
+* #### `ReportGetHandler`
+
+    ###### synopsis
+    Returns the report of a recording that was saved to persistent storage.
+
+    ###### request
+    `GET /api/v1/reports/:recordingName`
+
+    `recordingName` - The name of the recording to get the report for.
+
+    ###### response
+    `200` - The body is the requested report.
+
+    `401` - User authentication failed. The body is an error message.
+
+    `404` - The report could not be found. The body is an error message.
+
+    `407` - JMX authentication failed. The body is an error message.
+
+    `500` - There was an unexpected error. The body is an error message.
+
+    ###### example
+    ```
+    $ curl localhost:8181/api/v1/reports/localhost_foo_20200911T144545Z.jfr --output report.html
+      % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
+                                     Dload  Upload   Total   Spent    Left  Speed
+    100  116k  100  116k    0     0   134k      0 --:--:-- --:--:-- --:--:--  134k
+    ```
+
+
+* #### `TargetEventsGetHandler`
+
+    ###### synopsis
+    Returns a list of event types that can be produced by a target JVM.
+
+    ###### request
+    `GET /api/v1/targets/:targetId/events`
+
+    `targetId` - The location of the target JVM to connect to,
+    in the form of a `service:rmi:jmx://` JMX Service URL, or `hostname:port`.
+
+    ###### response
+    `200` - The body consists of a comma-seperated list of event types,
+    enclosed in square brackets.
+
+    The format for an event type is
+    `{"name":"$NAME","typeId":"$TYPE_ID","description":"$DESCRIPTION",
+    "category":[$CATEGORIES],"options":{$OPTIONS}}`
+
+    So a list with no events looks like `[]`,
+    a list with one event looks like `[{...}]`,
+    and a list with two events looks like `[{...},{...}]`.
+
+    `401` - User authentication failed. The body is an error message.
+
+    `404` - The target could not be found. The body is an error message.
+
+    `407` - JMX authentication failed. The body is an error message.
+
+    `500` - There was an unexpected error. The body is an error message.
+
+    ###### example
+    ```
+    $ curl localhost:8181/api/v1/targets/localhost/events --output events
+      % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
+                                     Dload  Upload   Total   Spent    Left  Speed
+    100 47013  100 47013    0     0   866k      0 --:--:-- --:--:-- --:--:--  866k
+    ```
+
+
+* #### `TargetRecordingDeleteHandler`
+
+    ###### synopsis
+    Deletes a recording from a target JVM, removing it from the target JVM
+    and freeing the buffer memory used. The recording will be stopped if
+    it is running at the time of deletion.
+
+    ###### request
+    `DELETE /api/v1/targets/:targetId/recordings/:recordingName`
+
+    `targetId` - The location of the target JVM to connect to,
+    in the form of a `service:rmi:jmx://` JMX Service URL, or `hostname:port`.
```

Comment:
This applies to all paths with `:targetId` and `:recordingName` - just to be safe, perhaps we should specify that these path components should be URI-encoded? Otherwise the resulting string-interpolated URL may be invalid or may be interpreted differently than the client intended.

Reply:
Could you explain a bit what URI-encoded means? I wasn't able to look that up.

Reply:
https://www.w3schools.com/tags/ref_urlencode.ASP / https://en.wikipedia.org/wiki/Percent-encoding

Reply:
ie clients should not try to request ex. `DELETE /api/v1/targets/service:jmx:rmi:///jndi/rmi://container-jfr:9091/jmxrmi/recordings/foo`, but rather `DELETE /api/v1/targets/service%3Ajmx%3Armi%3A%2F%2F%2Fjndi%2Frmi%3A%2F%2Fcontainer-jfr%3A9091%2Fjmxrmi%2Frecordings%2Ffoo/recordings/foo`